src/connector/src/sink/kafka.rs

```diff
@@ -395,36 +401,51 @@ impl KafkaSinkWriter {
 
     /// The actual `send_result` function, will be called when the `KafkaSinkWriter` needs to sink
     /// messages
-    async fn send_result<'a, K, P>(&'a self, mut record: FutureRecord<'a, K, P>) -> KafkaResult<()>
+    async fn send_result<'a, K, P>(
+        &'a mut self,
+        mut record: FutureRecord<'a, K, P>,
+    ) -> KafkaResult<()>
     where
         K: ToBytes + ?Sized,
         P: ToBytes + ?Sized,
     {
         // The error to be returned
         let mut err = KafkaError::Canceled;
 
+        // First take the ownership of the exist buffer
```

Comment:
The logic here is too complicated.

It can be as simple as 
```
if self.future_buffer.len() == self.max_limit {
    // wait for some future(not all) in future_buffer to finish to release the quota.
} else {
    let future = self.send_result.inner(record);
    self.future_buffer.push(future);
}
```

For the step that waits for some future to finish, you can use `VecDeque` instead of `Vec` to buffer the future, and always push to the back, and pop from front and then await on the future at the queue head. You can also investigate on `FuturesUnordered` or `FuturesOrdered`.

Reply:
The current approach is:

1. Check if the buffer is full; if so, await all the current futures (which is `self.commit`)
2. Otherwise, push the future into the buffer and return.
3. If the enqueue buffer is full, retry til the preset limit.

I think this is reasonable, since we've already wrapped the delivered process into `commit` per 10 futures, and based on the current 10 futures limit, I think it's also efficient just to wait the 10 buffered futures to be delivered and then begin the next turn.
Ultimately if ever encounters error, we will change the status back to the latest checkpoint.

Reply:
The logic makes some sense when the limit is 10, a relatively small value. If we decide to change to a larger values in the future, such 65536, it makes no sense to still wait for all futures to finish.

Besides, it is discouraged to call `commit` inside `send_result`, since the `commit` method is one of the methods in the `SinkWriter` trait, and its semantic is to be called on a externally triggered checkpoint, while in `send_result` this is the not the case. If we really want to reuse some logic, we can have a separate method named like `ensure_delivery`, and call it in both `send_result` and `commit`.

Also in current code I am confused of the usage of `commit_flag` and `push_flag`. It seems that the logic can be implemented without taking out self.future_delivery_buffer and using the two flags.

Reply:
The `commit` I mentioned here is not actually the `SinkWriter::commit`, instead a wrapper function of `KafkaSinkWriter::commit` that will clear the current `future_delivery_buffer` and ensure delivery. (And it's the function discussed below, the one to change to `try_join_all`. The semantic of which is essentially the same as the `ensure_delivery` wrapper function mentioned above)
For the use of `commit_flag` and `push_flag`, the former one is to check if the size is at the limit, the latter one is to check if this call has successfully sent the data and push the future into the `future_record_buffer`, then exchange the local `future_buffer` and the one in `KafkaSinkWriter` using `std::mem::swap`, in order to make borrow checker happy.
And yes, for a rather large buffer value such as 65536, we could actually adapt to a non-blocking approach.
I've tried implementing the non-blocking logic but it seems tricky and ambiguous (i.e., To wait for the release of quota, we will typically await the future at the front of the deque, but should we launch a background thread as the consumer to keep awaiting whenever the buffer is not empty, or some other approach to consume the buffer?), so may be future reviewed & discussed.

Reply:
Do we really need any non-blocking approach? The reason for non-blocking approach is that while waiting for the message being delivered, we have something else to do and we don't want to be blocked. However, when the buffer is full, we have nothing else to do but only wait for the release of quota. If so, we only need to take the future at the queue front, and simply await on it. 

Actually, the current functionality provided by the rdkafka client is already non-blocking, whose implementation allows adding records to its inner queue and asynchronously delivering message to external kafka, and at the end notifying the delivery. We don't need to implement any extra non-blocking logic.

Besides, the two flags can be avoided. We can in-place update any states in where the flags are set now, instead of setting flags and applied the update afterward. If the borrow checker reports any error, we can resolve it together.